File: internal/export/export.go
package export

import (
	"context"
	"encoding/csv"
	"fmt"
	"math/big"
	"os"
	"strings"
	"time"

	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgconn"
	"github.com/jackc/pgx/v5/pgtype"
	"github.com/parquet-go/parquet-go"
	"github.com/parquet-go/parquet-go/compress/zstd"

	"github.com/louloulibs/wrds-download/internal/db"
)

// Options controls the export behaviour.
type Options struct {
	Format string // "parquet" or "csv"
}

const rowGroupSize = 10_000

// Export runs query against the WRDS PostgreSQL instance and writes output to outPath.
// Format is determined by opts.Format (default: parquet).
func Export(query, outPath string, opts Options) error {
	format := strings.ToLower(opts.Format)
	if format == "" {
		if strings.HasSuffix(strings.ToLower(outPath), ".csv") {
			format = "csv"
		} else {
			format = "parquet"
		}
	}

	dsn, err := db.DSNFromEnv()
	if err != nil {
		return fmt.Errorf("dsn: %w", err)
	}

	ctx := context.Background()
	conn, err := pgx.Connect(ctx, dsn)
	if err != nil {
		return fmt.Errorf("connect: %w", err)
	}
	defer conn.Close(ctx)

	rows, err := conn.Query(ctx, query)
	if err != nil {
		return fmt.Errorf("query: %w", err)
	}
	defer rows.Close()

	switch format {
	case "csv":
		return writeCSV(rows, outPath)
	default:
		return writeParquet(rows, outPath)
	}
}

// writeCSV streams rows into a CSV file with a header row.
func writeCSV(rows pgx.Rows, outPath string) error {
	f, err := os.Create(outPath)
	if err != nil {
		return fmt.Errorf("create csv: %w", err)
	}
	defer f.Close()

	w := csv.NewWriter(f)
	defer w.Flush()

	fds := rows.FieldDescriptions()
	header := make([]string, len(fds))
	for i, fd := range fds {
		header[i] = fd.Name
	}
	if err := w.Write(header); err != nil {
		return fmt.Errorf("write header: %w", err)
	}

	record := make([]string, len(fds))
	for rows.Next() {
		vals, err := rows.Values()
		if err != nil {
			return fmt.Errorf("scan row: %w", err)
		}
		for i, v := range vals {
			record[i] = formatValue(v)
		}
		if err := w.Write(record); err != nil {
			return fmt.Errorf("write row: %w", err)
		}
	}
	if err := rows.Err(); err != nil {
		return fmt.Errorf("rows: %w", err)
	}

	w.Flush()
	return w.Error()
}

// writeParquet streams rows into a Parquet file using parquet-go.
func writeParquet(rows pgx.Rows, outPath string) error {
	fds := rows.FieldDescriptions()

	schema, colTypes := buildParquetSchema(fds)

	f, err := os.Create(outPath)
	if err != nil {
		return fmt.Errorf("create parquet: %w", err)
	}
	defer f.Close()

	writer := parquet.NewGenericWriter[map[string]any](f,
		schema,
		parquet.Compression(&zstd.Codec{}),
		parquet.DefaultEncodingFor(parquet.ByteArray, &parquet.Plain),
	)

	buf := make([]map[string]any, 0, rowGroupSize)

	for rows.Next() {
		vals, err := rows.Values()
		if err != nil {
			return fmt.Errorf("scan row: %w", err)
		}

		row := make(map[string]any, len(fds))
		for i, v := range vals {
			row[fds[i].Name] = convertValue(v, colTypes[i])
		}
		buf = append(buf, row)

		if len(buf) >= rowGroupSize {
			if _, err := writer.Write(buf); err != nil {
				return fmt.Errorf("write row group: %w", err)
			}
			buf = buf[:0]
		}
	}
	if err := rows.Err(); err != nil {
		return fmt.Errorf("rows: %w", err)
	}

	// Flush remaining rows.
	if len(buf) > 0 {
		if _, err := writer.Write(buf); err != nil {
			return fmt.Errorf("write final rows: %w", err)
		}
	}

	return writer.Close()
}

// colType tags how we convert PG values for Parquet.
type colType int

const (
	colString colType = iota
	colBool
	colInt32
	colInt64
	colFloat32
	colFloat64
	colDate      // days since epoch → int32
	colTimestamp // microseconds since epoch → int64
)

// buildParquetSchema maps PG field descriptors to a parquet schema.
func buildParquetSchema(fds []pgconn.FieldDescription) (*parquet.Schema, []colType) {
	cols := make([]colType, len(fds))
	group := make(parquet.Group, len(fds))

	for i, fd := range fds {
		var node parquet.Node

		switch fd.DataTypeOID {
		case 16: // bool
			cols[i] = colBool
			node = parquet.Optional(parquet.Leaf(parquet.BooleanType))
		case 21: // int2
			cols[i] = colInt32
			node = parquet.Optional(parquet.Leaf(parquet.Int32Type))
		case 23: // int4
			cols[i] = colInt32
			node = parquet.Optional(parquet.Leaf(parquet.Int32Type))
		case 20: // int8
			cols[i] = colInt64
			node = parquet.Optional(parquet.Leaf(parquet.Int64Type))
		case 700: // float4
			cols[i] = colFloat32
			node = parquet.Optional(parquet.Leaf(parquet.FloatType))
		case 701: // float8
			cols[i] = colFloat64
			node = parquet.Optional(parquet.Leaf(parquet.DoubleType))
		case 1082: // date
			cols[i] = colDate
			node = parquet.Optional(parquet.Date())
		case 1114, 1184: // timestamp, timestamptz
			cols[i] = colTimestamp
			node = parquet.Optional(parquet.Timestamp(parquet.Microsecond))
		default:
			// text (25), varchar (1043), char (18, 1042), numeric (1700), etc.
			cols[i] = colString
			node = parquet.Optional(parquet.String())
		}

		group[fd.Name] = node
	}

	return parquet.NewSchema("wrds", group), cols
}

var epoch = time.Date(1970, 1, 1, 0, 0, 0, 0, time.UTC)

// convertValue converts a pgx-scanned value to the appropriate Go type for parquet-go.
func convertValue(v any, ct colType) any {
	if v == nil {
		return nil
	}

	switch ct {
	case colBool:
		if b, ok := v.(bool); ok {
			return b
		}
	case colInt32:
		switch n := v.(type) {
		case int16:
			return int32(n)
		case int32:
			return n
		case int64:
			return int32(n)
		}
	case colInt64:
		switch n := v.(type) {
		case int64:
			return n
		case int32:
			return int64(n)
		case int16:
			return int64(n)
		}
	case colFloat32:
		if f, ok := v.(float32); ok {
			return f
		}
		if f, ok := v.(float64); ok {
			return float32(f)
		}
	case colFloat64:
		if f, ok := v.(float64); ok {
			return f
		}
		if f, ok := v.(float32); ok {
			return float64(f)
		}
	case colDate:
		if t, ok := v.(time.Time); ok {
			days := int32(t.Sub(epoch).Hours() / 24)
			return days
		}
	case colTimestamp:
		if t, ok := v.(time.Time); ok {
			return t.Sub(epoch).Microseconds()
		}
	case colString:
		return formatValue(v)
	}

	// Fallback: stringify.
	return formatValue(v)
}

// formatValue converts any value to its string representation.
func formatValue(v any) string {
	if v == nil {
		return ""
	}
	switch val := v.(type) {
	case string:
		return val
	case []byte:
		return string(val)
	case time.Time:
		if val.Hour() == 0 && val.Minute() == 0 && val.Second() == 0 && val.Nanosecond() == 0 {
			return val.Format("2006-01-02")
		}
		return val.Format(time.RFC3339)
	case pgtype.Numeric:
		if !val.Valid {
			return ""
		}
		if val.NaN {
			return "NaN"
		}
		if val.InfinityModifier == pgtype.Infinity {
			return "Infinity"
		}
		if val.InfinityModifier == pgtype.NegativeInfinity {
			return "-Infinity"
		}
		// Convert to big.Float for string representation.
		bi := val.Int
		if bi == nil {
			bi = new(big.Int)
		}
		bf := new(big.Float).SetInt(bi)
		if val.Exp < 0 {
			divisor := new(big.Float).SetInt(new(big.Int).Exp(big.NewInt(10), big.NewInt(int64(-val.Exp)), nil))
			bf.Quo(bf, divisor)
		} else if val.Exp > 0 {
			multiplier := new(big.Float).SetInt(new(big.Int).Exp(big.NewInt(10), big.NewInt(int64(val.Exp)), nil))
			bf.Mul(bf, multiplier)
		}
		return bf.Text('f', -1)
	default:
		return fmt.Sprintf("%v", val)
	}
}

File: python/src/wrds_dl/export.py
"""Export query results to Parquet or CSV with streaming and progress."""

from __future__ import annotations

import csv
from decimal import Decimal
from typing import Callable

import psycopg
import pyarrow as pa
import pyarrow.parquet as pq

from wrds_dl.db import dsn_from_env

ROW_GROUP_SIZE = 10_000

# Map PostgreSQL type OIDs to PyArrow types.
_PG_OID_TO_ARROW: dict[int, pa.DataType] = {
    16: pa.bool_(),        # bool
    21: pa.int32(),        # int2
    23: pa.int32(),        # int4
    20: pa.int64(),        # int8
    700: pa.float32(),     # float4
    701: pa.float64(),     # float8
    1082: pa.date32(),     # date
    1114: pa.timestamp("us"),  # timestamp
    1184: pa.timestamp("us", tz="UTC"),  # timestamptz
}


def _arrow_type_for_oid(oid: int) -> pa.DataType:
    return _PG_OID_TO_ARROW.get(oid, pa.string())


def export_data(
    query: str,
    out_path: str,
    fmt: str = "parquet",
    progress_fn: Callable[[int], None] | None = None,
) -> None:
    """Run *query* against WRDS and write results to *out_path*."""
    conn = psycopg.connect(dsn_from_env())
    try:
        with conn.cursor(name="wrds_export") as cur:
            cur.itersize = ROW_GROUP_SIZE
            cur.execute(query)

            if cur.description is None:
                raise RuntimeError("Query returned no columns")

            col_names = [desc.name for desc in cur.description]
            col_oids = [desc.type_code for desc in cur.description]

            if fmt == "csv":
                _write_csv(cur, col_names, out_path, progress_fn)
            else:
                _write_parquet(cur, col_names, col_oids, out_path, progress_fn)
    finally:
        conn.close()


def _write_csv(
    cur: psycopg.Cursor,
    col_names: list[str],
    out_path: str,
    progress_fn: Callable[[int], None] | None,
) -> None:
    with open(out_path, "w", newline="") as f:
        writer = csv.writer(f)
        writer.writerow(col_names)
        total = 0
        for row in cur:
            writer.writerow(_format_row(row))
            total += 1
            if progress_fn and total % ROW_GROUP_SIZE == 0:
                progress_fn(total)


def _write_parquet(
    cur: psycopg.Cursor,
    col_names: list[str],
    col_oids: list[int],
    out_path: str,
    progress_fn: Callable[[int], None] | None,
) -> None:
    arrow_types = [_arrow_type_for_oid(oid) for oid in col_oids]
    schema = pa.schema([(name, typ) for name, typ in zip(col_names, arrow_types)])

    writer = pq.ParquetWriter(out_path, schema, compression="zstd")
    try:
        batch_rows: list[tuple] = []
        total = 0

        for row in cur:
            batch_rows.append(row)
            if len(batch_rows) >= ROW_GROUP_SIZE:
                _flush_batch(writer, schema, batch_rows, col_names)
                total += len(batch_rows)
                batch_rows = []
                if progress_fn:
                    progress_fn(total)

        if batch_rows:
            _flush_batch(writer, schema, batch_rows, col_names)
            total += len(batch_rows)
    finally:
        writer.close()


def _flush_batch(
    writer: pq.ParquetWriter,
    schema: pa.Schema,
    rows: list[tuple],
    col_names: list[str],
) -> None:
    """Convert a batch of rows into a PyArrow table and write it."""
    columns: dict[str, list] = {name: [] for name in col_names}
    for row in rows:
        for i, val in enumerate(row):
            # Strip trailing zeros from Decimal values (numeric columns)
            # so output matches Go's pgx behaviour.
            if isinstance(val, Decimal):
                val = str(val.normalize())
            columns[col_names[i]].append(val)

    arrays = []
    for i, name in enumerate(col_names):
        try:
            arrays.append(pa.array(columns[name], type=schema.field(name).type))
        except (pa.ArrowInvalid, pa.ArrowTypeError):
            # Fallback: convert to strings
            arrays.append(pa.array([str(v) if v is not None else None for v in columns[name]],
                                   type=pa.string()))

    table = pa.table(dict(zip(col_names, arrays)))
    writer.write_table(table)


def _format_row(row: tuple) -> list[str]:
    """Format a row for CSV output."""
    out = []
    for v in row:
        if v is None:
            out.append("")
        elif isinstance(v, Decimal):
            out.append(str(v.normalize()))
        else:
            out.append(str(v))
    return out
